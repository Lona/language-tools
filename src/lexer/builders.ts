import {
  Action,
  Rule,
  Position,
  NextAction,
  PushAction,
  PopAction,
  StateDefinition,
  Token,
} from './types'

/* -------- Actions -------- */

export function nextAction(value: string): NextAction {
  return { type: 'next', value }
}

export function pushAction(value: string): PushAction {
  return { type: 'push', value }
}

export function pop(): PopAction {
  return { type: 'pop' }
}

/* -------- Rules -------- */

type RuleOptions = {
  discard?: boolean
  action?: Action
  // print?: TokenPrintPattern
}

/**
 * Match a keyword followed by a word boundary
 */
export function keyword(name: string): Rule {
  return rule(name, new RegExp(name + '\\b'))
}

export function rule(name: string): Rule
export function rule(name: string, pattern: string | RegExp): Rule
export function rule(
  name: string,
  pattern: string | RegExp,
  options: RuleOptions
): Rule
export function rule(
  name: string,
  pattern?: string | RegExp,
  options: RuleOptions = {}
): Rule {
  const source = pattern instanceof RegExp ? pattern.source : pattern

  return {
    name,
    pattern: source ?? name,
    discard: options.discard ?? false,
    // print: options.print ?? { type: 'literal', value: pattern },
    action: options.action,
  }
}

/* -------- Tokens -------- */

type TokenOptions = { value?: string; values?: string[]; position?: Position }

export function token(type: string): Token
export function token(type: string, value: string): Token
export function token(type: string, options: TokenOptions): Token

export function token(
  type: string,
  optionsOrValue?: TokenOptions | string
): Token {
  const options =
    typeof optionsOrValue === 'undefined'
      ? {}
      : typeof optionsOrValue === 'string'
      ? { values: [optionsOrValue] }
      : optionsOrValue

  return {
    type,
    values: options.values ?? [],
    position: options.position ?? { start: 0, end: 0 },
  }
}

/* -------- States -------- */

export type RuleShorthand =
  | string
  | [string]
  | [string, string | RegExp]
  | [string, string | RegExp, string]
  | [string, string | RegExp, RuleOptions]

/**
 * Support Monarch lexer shorthand patterns
 */
export function normalizeRule(rule: Rule | RuleShorthand): Rule {
  let arrayOrRule = typeof rule === 'string' ? [rule] : rule

  if (arrayOrRule instanceof Array) {
    const [name, pattern, options] = arrayOrRule

    const source = pattern instanceof RegExp ? pattern.source : pattern

    return {
      name,
      pattern: source ?? name,
      discard: false,
      ...(typeof options === 'string'
        ? { action: { type: 'next', value: options } }
        : options),
    }
  }

  return arrayOrRule
}

export function state(
  name: string,
  rules: Parameters<typeof normalizeRule>[0][]
): StateDefinition {
  if (rules.length === 0) {
    throw new Error('A lexer state must contain at least 1 rule')
  }

  return { name, rules: rules.map(normalizeRule) }
}
