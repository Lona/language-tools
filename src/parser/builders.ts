import {
  ConsumePattern,
  SequencePattern,
  OrPattern,
  ManyPattern,
  OptionPattern,
  Pattern,
  PrintLiteralPattern,
  PrintLinePattern,
  PrintIndentPattern,
  PrintGroupPattern,
  ThunkPattern,
  PrintIfPattern,
  TerminalPattern,
  LineKind,
} from './types'
import { ParserPattern } from './Parser'

let initialId = 0

// Exported for testing
export function resetId() {
  initialId = 0
}

function id(): string {
  return String(initialId++)
}

type LabelOptions = { as?: string; label?: string }

type PatternShorthand<T> =
  | Pattern<T>
  | PatternShorthand<T>[]
  | (() => Pattern<T>)

function normalizePattern<T>(pattern: PatternShorthand<T>): Pattern<T> {
  if (typeof pattern === 'function') {
    return thunk(pattern)
  } else if (pattern instanceof Array) {
    return sequence(pattern)
  }

  return pattern
}

export function consume<T>(
  value: T,
  options: LabelOptions | string = {}
): ConsumePattern<T> {
  const normalizedOptions =
    typeof options === 'string' ? { as: options } : options
  return { id: id(), type: 'consume', value, ...normalizedOptions }
}

export function sequence<T>(
  value: PatternShorthand<T>[],
  options?: { select?: string } & LabelOptions
): SequencePattern<T> {
  return {
    id: id(),
    type: 'sequence',
    value: value.map(normalizePattern),
    ...options,
  }
}

export function or<T>(
  value: PatternShorthand<T>[],
  options: LabelOptions & { typeNames?: string[]; parallel?: boolean } = {}
): OrPattern<T> {
  if (options.typeNames && value.length !== options.typeNames.length) {
    throw new Error('typeNames length must match number of patterns')
  }

  return {
    id: id(),
    type: 'or',
    value: value.map(normalizePattern),
    parallel: options.parallel ?? false,
    ...options,
  }
}

export function typedOr<T>(
  value: [string, PatternShorthand<T>][],
  options: LabelOptions & { parallel?: boolean } = {}
): OrPattern<T> {
  return {
    id: id(),
    type: 'or',
    value: value.map(([_, pattern]) => normalizePattern(pattern)),
    typeNames: value.map(([name]) => name),
    parallel: options.parallel ?? false,
    ...options,
  }
}

export function many<T>(
  value: PatternShorthand<T>,
  options: { separator?: PatternShorthand<T> } & LabelOptions = {}
): ManyPattern<T> {
  const { separator, ...rest } = options

  return {
    id: id(),
    type: 'many',
    value: normalizePattern(value),
    ...(separator && { separator: normalizePattern(separator) }),
    ...rest,
  }
}

export function option<T>(
  value: PatternShorthand<T>,
  options:
    | ({ select?: string; defaultValue?: unknown } & LabelOptions)
    | string = {}
): OptionPattern<T> {
  const normalizedOptions =
    typeof options === 'string' ? { as: options } : options
  return {
    id: id(),
    type: 'option',
    value: normalizePattern(value),
    ...normalizedOptions,
  }
}

export function thunk<T>(
  value: () => Pattern<T>,
  options: LabelOptions | string = {}
): ThunkPattern<T> {
  const normalizedOptions =
    typeof options === 'string' ? { as: options } : options
  return { id: id(), type: 'thunk', value, ...normalizedOptions }
}

// Printing

export function printLiteral<T>(value: string): PrintLiteralPattern<T> {
  return { id: id(), type: 'printLiteral', value }
}

export function printLine<T>(kind?: 'soft' | 'hard'): PrintLinePattern<T> {
  return { id: id(), type: 'printLine', ...(kind && { kind }) }
}

export function printIndent<T>(
  value: PatternShorthand<T>
): PrintIndentPattern<T> {
  return { id: id(), type: 'printIndent', value: normalizePattern(value) }
}

export function printGroup<T>(
  value: PatternShorthand<T>
): PrintGroupPattern<T> {
  return { id: id(), type: 'printGroup', value: normalizePattern(value) }
}

export function printIf<T>(
  value: PatternShorthand<T>,
  {
    before,
    after,
    reference,
  }: {
    reference: string
    before?: TerminalPattern<T>
    after?: TerminalPattern<T>
  }
): PrintIfPattern<T> {
  return {
    id: id(),
    type: 'printIf',
    value: normalizePattern(value),
    reference,
    ...(before && { before }),
    ...(after && { after }),
  }
}

export function printIndentedMany<T>(
  value: ManyPattern<T>,
  { padding }: { padding?: LineKind } = {}
) {
  if (!value.as) {
    throw new Error(`The 'printIndentedMany' helper requires an 'as' label`)
  }

  return printGroup(
    printIf(
      printIndent(
        printIf(value, { reference: value.as, before: printLine(padding) })
      ),
      { reference: value.as, after: printLine(padding) }
    )
  )
}

export function language(patterns: {
  [key: string]: PatternShorthand<string>
}): { [key: string]: ParserPattern } {
  return Object.fromEntries(
    Object.entries(patterns).map(([key, pattern]) => [
      key,
      normalizePattern(pattern),
    ])
  )
}
