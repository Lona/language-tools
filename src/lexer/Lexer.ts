import { inspect } from 'util'
import memoize from '../utils/memoize'
import { RuleShorthand, state } from './builders'
import { Rule, StateDefinition, Token } from './types'

const getRegExp = memoize(
  (pattern: string): RegExp => new RegExp('^(?:' + pattern + ')')
)

export class Lexer {
  stateDefinitions: StateDefinition[]

  static stateless(rules: (Rule | RuleShorthand)[]): Lexer {
    return new Lexer([state('main', rules)])
  }

  constructor(states: StateDefinition[]) {
    if (states.length === 0) {
      throw new Error('A lexer must be defined with at least one state')
    }

    this.stateDefinitions = states
  }

  tokenize(
    source: string,
    initialState?: string[],
    { positionTracking }: { positionTracking?: boolean } = {}
  ): Token[] {
    const stack: string[] = initialState ?? [this.stateDefinitions[0].name]

    for (let name of stack) {
      if (!this.stateDefinitions.some((state) => state.name === name)) {
        throw new Error(`Lexer state "${name}" isn't defined`)
      }
    }

    let length = source.length
    let pos = 0
    let tokens: Token[] = []

    main: while (pos < length) {
      const currentState = stack[stack.length - 1]

      const stateDefinition = this.stateDefinitions.find(
        (definition) => definition.name === currentState
      )

      if (!stateDefinition) {
        console.log(currentState, stack)
        throw new Error(`Invalid lexer state: ${currentState}.`)
      }

      const { rules } = stateDefinition

      const slice = source.slice(pos)

      for (let rule of rules) {
        let regExp = getRegExp(rule.pattern)

        const match = slice.match(regExp)

        if (match) {
          const [value, ...values] = match

          if (value.length === 0 && !rule.action) {
            throw new Error(
              `Lexer stalled: the pattern matched, but no input was consumed and no action taken for rule: ${inspect(
                rule
              )}.`
            )
          }

          if (rule.discard !== true) {
            tokens.push({
              type: rule.name,
              values,
              position: {
                start: positionTracking !== false ? pos : 0,
                end: positionTracking !== false ? pos + value.length : 0,
              },
            })
          }

          if (rule.action) {
            switch (rule.action.type) {
              case 'next':
                stack[stack.length - 1] = rule.action.value
                break
              case 'push':
                stack.push(rule.action.value)
                break
              case 'pop':
                if (stack.length === 1) {
                  throw new Error(
                    `Lexer tried to pop to empty state from state: ${stack[0]}`
                  )
                }

                stack.pop()
                break
            }
          }

          pos += value.length

          continue main
        }
      }

      console.error(
        `Failed to parse: ${slice}\n`,
        currentState,
        rules.map((rule) => rule.pattern)
      )

      break
    }

    return tokens
  }
}
