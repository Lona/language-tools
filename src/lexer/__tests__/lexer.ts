import { rule, state } from '../builders'
import { Lexer } from '../Lexer'

it('tokenizes simple tokens', () => {
  const lexer = Lexer.stateless([
    ['from'],
    ['select'],
    ['where'],
    ['identifier', '[a-zA-Z]\\w*'],
    ['integer', '0|[1-9]\\d*'],
    ['_', '\\W+'],
  ])

  const tokens = lexer.tokenize('select foo from 123')

  expect(tokens).toMatchSnapshot()
})

it('captures groups', () => {
  const lexer = Lexer.stateless([['tag', '<(\\w+)>']])

  const tokens = lexer.tokenize('<hi>')

  expect(tokens).toMatchSnapshot()
})

it('supports states', () => {
  const lexer = new Lexer([
    state('main', [
      ['quote', '"', 'string'],
      ['content', '[^"]+'],
    ]),
    state('string', [
      ['quote', '"', 'main'],
      ['string', '[^"]+'],
    ]),
  ])

  const tokens = lexer.tokenize('hello world "this is a string" ok')

  expect(tokens).toMatchSnapshot()
})

it('fails if no input is consumed', () => {
  const lexer = Lexer.stateless([rule('nothing', '')])

  expect(() => {
    lexer.tokenize('hello world "this is a string" ok')
  }).toThrow()
})
