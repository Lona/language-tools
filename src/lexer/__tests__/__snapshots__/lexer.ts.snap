// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`captures groups 1`] = `
Array [
  Object {
    "position": Object {
      "end": 4,
      "start": 0,
    },
    "type": "tag",
    "values": Array [
      "hi",
    ],
  },
]
`;

exports[`supports states 1`] = `
Array [
  Object {
    "position": Object {
      "end": 12,
      "start": 0,
    },
    "type": "content",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 13,
      "start": 12,
    },
    "type": "quote",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 29,
      "start": 13,
    },
    "type": "string",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 30,
      "start": 29,
    },
    "type": "quote",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 33,
      "start": 30,
    },
    "type": "content",
    "values": Array [],
  },
]
`;

exports[`tokenizes simple tokens 1`] = `
Array [
  Object {
    "position": Object {
      "end": 6,
      "start": 0,
    },
    "type": "select",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 7,
      "start": 6,
    },
    "type": "_",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 10,
      "start": 7,
    },
    "type": "identifier",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 11,
      "start": 10,
    },
    "type": "_",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 15,
      "start": 11,
    },
    "type": "from",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 16,
      "start": 15,
    },
    "type": "_",
    "values": Array [],
  },
  Object {
    "position": Object {
      "end": 19,
      "start": 16,
    },
    "type": "integer",
    "values": Array [],
  },
]
`;
